src/rgw/rgw_rest_sts.cc

```diff
@@ -223,137 +224,264 @@ WebTokenEngine::get_cert_url(const string& iss, const DoutPrefixProvider *dpp, o
+      certs.begin(), certs.end(), pattern);
+      it != sregex_iterator(); it++) {
+      smatch match;
+      match = *it;
+      x5c.emplace_back(match.str(1));
+      ldpp_dout(dpp, 10) << "Matched: " << match.str(1) << dendl;
+  }
+  return x5c;
+}
+
+void
+WebTokenEngine::validate_signature_using_cert(const DoutPrefixProvider* dpp, const jwt::decoded_jwt& decoded, const string& algorithm, const vector<string>& certs, const vector<string>& thumbprints, bool add_pem_str) const
+{
+  string cert;
+  bool found_valid_cert = false;
+  for (auto& it : certs) {
+    if (add_pem_str) {
+      cert = "-----BEGIN CERTIFICATE-----\n" + it + "\n-----END CERTIFICATE-----";
+    } else {
+      cert = it;
+    }
+    if (is_cert_valid(thumbprints, cert)) {
+      found_valid_cert = true;
+      break;
+    }
+  }
+  if (! found_valid_cert) {
+    ldpp_dout(dpp, 0) << "Cert doesn't match that with the thumbprints registered with oidc provider: " << cert.c_str() << dendl;
+    throw -EINVAL;
+  }
+  cert.clear();
+  if (add_pem_str) {
+    cert = "-----BEGIN CERTIFICATE-----\n" + certs[0] + "\n-----END CERTIFICATE-----"; //first cert always contains the public key
+  } else {
+    cert = certs[0];
+  }
+  ldpp_dout(dpp, 20) << "Certificate is: " << cert.c_str() << dendl;
+  //verify method takes care of expired tokens also
+  try {
+    if (algorithm == "RS256") {
+      auto verifier = jwt::verify()
+                  .allow_algorithm(jwt::algorithm::rs256{cert});
+
+      verifier.verify(decoded);
+    } else if (algorithm == "RS384") {
+      auto verifier = jwt::verify()
+                  .allow_algorithm(jwt::algorithm::rs384{cert});
+
+      verifier.verify(decoded);
+    } else if (algorithm == "RS512") {
+      auto verifier = jwt::verify()
+                  .allow_algorithm(jwt::algorithm::rs512{cert});
+
+      verifier.verify(decoded);
+    } else if (algorithm == "ES256") {
+      auto verifier = jwt::verify()
+                  .allow_algorithm(jwt::algorithm::es256{cert});
+
+      verifier.verify(decoded);
+    } else if (algorithm == "ES384") {
+      auto verifier = jwt::verify()
+                  .allow_algorithm(jwt::algorithm::es384{cert});
+
+      verifier.verify(decoded);
+    } else if (algorithm == "ES512") {
+      auto verifier = jwt::verify()
+                    .allow_algorithm(jwt::algorithm::es512{cert});
+
+      verifier.verify(decoded);
+    } else if (algorithm == "PS256") {
+      auto verifier = jwt::verify()
+                    .allow_algorithm(jwt::algorithm::ps256{cert});
+
+      verifier.verify(decoded);
+    } else if (algorithm == "PS384") {
+      auto verifier = jwt::verify()
+                    .allow_algorithm(jwt::algorithm::ps384{cert});
+
+      verifier.verify(decoded);
+    } else if (algorithm == "PS512") {
+      auto verifier = jwt::verify()
+                    .allow_algorithm(jwt::algorithm::ps512{cert});
+
+      verifier.verify(decoded);
+    }
+  } catch (std::runtime_error& e) {
```

Comment:
Why std::runtime_error here instead of std::exception?

Also catch by const reference `(const std::runtime_error& e)`. A bit of a nitpick. At one time it was thought to increase exception performance but on modern compilers it doesn't seem to make a difference.